src/class_singleCleaner.py

```diff
@@ -115,54 +114,28 @@ def run(self):
-                                continue
-                            # scrap old nodes with low rating
-                            if now - knownnodes.knownNodes[stream][node]["lastseen"] > 10800 and knownnodes.knownNodes[stream][node]["rating"] <= knownnodes.knownNodesForgetRating:
-                                shared.needToWriteKnownNodesToDisk = True
-                                del knownnodes.knownNodes[stream][node]
-                                continue
-                        except TypeError:
-                            print "Error in %s" % node
-                    keys = []
-
-            # Let us write out the knowNodes to disk
-            # if there is anything new to write out.
-            if shared.needToWriteKnownNodesToDisk:
-                try:
-                    knownnodes.saveKnownNodes()
-                except Exception as err:
-                    if "Errno 28" in str(err):
-                        logger.fatal(
-                            '(while receiveDataThread'
-                            ' knownnodes.needToWriteKnownNodesToDisk)'
-                            ' Alert: Your disk or data storage volume'
-                            ' is full. '
-                        )
-                        queues.UISignalQueue.put((
-                            'alert',
-                            (tr._translate("MainWindow", "Disk full"),
-                             tr._translate(
-                                 "MainWindow",
-                                 'Alert: Your disk or data storage volume'
-                                 ' is full. Bitmessage will now exit.'),
-                                True)
-                            ))
-                        # FIXME redundant?
-                        if shared.daemon or not state.enableGUI:
-                            os._exit(0)
-                shared.needToWriteKnownNodesToDisk = False
+            try:
+                # Cleanup knownnodes and handle possible severe exception
+                # while writing it to disk
+                knownnodes.cleanupKnownNodes()
+            except Exception as err:
+                if "Errno 28" in str(err):
+                    logger.fatal(
+                        '(while writing knownnodes to disk)'
+                        ' Alert: Your disk or data storage volume is full.'
+                    )
+                    queues.UISignalQueue.put((
+                        'alert',
+                        (tr._translate("MainWindow", "Disk full"),
+                         tr._translate(
+                             "MainWindow",
+                             'Alert: Your disk or data storage volume'
+                             ' is full. Bitmessage will now exit.'),
+                            True)
+                    ))
+                    # FIXME redundant?
+                    if shared.daemon or not state.enableGUI:
+                        sys.exit(1)
```

Comment:
The reason why `os._exit()` rather than `sys.exit()` was used here is to avoid calling `shutdown.doCleanShutdown()`, which may screw up things even more if there is low diskspace.